README.md
<!--
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.

For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages).

For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages).
-->

# Bilinear Gradient

## Features

This package will allow you to draw the `BilinearGradientDecorator` widget, as `LinearGradient` is already built in with Flutter.

![Example view](https://github.com/sthefanoss/bilinear-gradient/blob/82b7c98f8e19862597d6ac6e562ab7e694029abd/screenshots/example.png)

But don't confuse the `Container`'s `LinearGradient` with the `BilinearGradientDecorator`, since the former is an attribute of a `Container`'s decoration, while the latter is a stand alone `Widget`.

## Usage


```dart
import 'package:bilinear_gradient/bilinear_gradient.dart';
...

const topLeftColor = Color(0xFFF7F700);
const topRightColor = Color(0xFF00F700);
const bottomLeftColor = Color(0xFF7675C3);
const bottomRightColor = Color(0xFFF70075);

...
child: BilinearGradientDecorator(
        topLeftColor: topLeftColor,
        topRightColor: topRightColor,
        bottomLeftColor: bottomLeftColor,
        bottomRightColor: bottomRightColor,
        child: Center(
            child: DecoratedText('Bilinear Gradient')
        ),
    )
...
```

## Additional information
- Tested on Android/iOS/macOS with Flutter 3.7.0. Web did't worked on that version;
- Tested again in with Flutter 3.10.5. Everything work just fine at this time.


CHANGELOG.md
## 0.0.1+1
* fix `README.md` example.png path.

## 0.0.1
* Releasing package as poc.


lib/src/bilinear_painter.dart
part of '../bilinear_gradient.dart';

class _BilinearPainter extends CustomPainter {
  const _BilinearPainter({
    required this.shader,
    required this.topLeftColor,
    required this.topRightColor,
    required this.bottomLeftColor,
    required this.bottomRightColor,
  });

  final FragmentShader shader;
  final Color topLeftColor;
  final Color topRightColor;
  final Color bottomLeftColor;
  final Color bottomRightColor;

  @override
  void paint(Canvas canvas, Size size) {
    shader.setFloatUniforms((value) {
      value.setFloat(size.width);
      value.setFloat(size.height);
      value.setColor(topLeftColor);
      value.setColor(topRightColor);
      value.setColor(bottomLeftColor);
      value.setColor(bottomRightColor);
    });

    final paint = Paint()..shader = shader;
    canvas.drawRect(Rect.fromLTWH(0, 0, size.width, size.height), paint);
  }

  @override
  bool shouldRepaint(_BilinearPainter oldDelegate) {
    return topLeftColor != oldDelegate.topLeftColor ||
        topRightColor != oldDelegate.topRightColor ||
        bottomLeftColor != oldDelegate.bottomLeftColor ||
        bottomRightColor != oldDelegate.bottomRightColor;
  }
}


lib/src/bilinear_gradient_container.dart
part of '../bilinear_gradient.dart';

/// A container that allows you to implement 4 colors linear gradient.
class BilinearGradientDecorator extends StatelessWidget {
  const BilinearGradientDecorator({
    super.key,
    required this.topLeftColor,
    required this.topRightColor,
    required this.bottomLeftColor,
    required this.bottomRightColor,
    required this.child,
  });

  final Color topLeftColor;
  final Color topRightColor;
  final Color bottomLeftColor;
  final Color bottomRightColor;
  final Widget child;

  static final _assetKey = const AssetImage(
    'shaders/bilinear_gradient.frag',
    package: 'bilinear_gradient',
  ).keyName;

  @override
  Widget build(BuildContext context) {
    return ShaderBuilder(
      assetKey: _assetKey,
      (context, shader, child) => CustomPaint(
        painter: _BilinearPainter(
          shader: shader,
          topLeftColor: topLeftColor,
          topRightColor: topRightColor,
          bottomLeftColor: bottomLeftColor,
          bottomRightColor: bottomRightColor,
        ),
        child: this.child,
      ),
    );
  }
}


example/lib/main.dart
import 'package:bilinear_gradient/bilinear_gradient.dart';
import 'package:flutter/material.dart';

void main() => runApp(const MyApp());

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) => MaterialApp(
        theme: ThemeData.dark(useMaterial3: true),
        home: const HomePage(),
        title: 'Bilinear Gradient Example',
      );
}

class HomePage extends StatelessWidget {
  const HomePage({super.key});

  static const topLeftColor = Color(0xFFF7F700);
  static const topRightColor = Color(0xFF00F700);
  static const bottomLeftColor = Color(0xFF7675C3);
  static const bottomRightColor = Color(0xFFF70075);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: Text('Bilinear Gradient')),
      body: Stack(
        fit: StackFit.expand,
        children: [
          FractionallySizedBox(
            widthFactor: 0.1,
            heightFactor: 0.1,
            alignment: Alignment.topLeft,
            child: Container(
              color: topLeftColor,
              alignment: Alignment.center,
              child: DecoratedText('Yellow'),
            ),
          ),
          FractionallySizedBox(
            widthFactor: 0.1,
            heightFactor: 0.1,
            alignment: Alignment.topRight,
            child: Container(
              color: topRightColor,
              alignment: Alignment.center,
              child: DecoratedText('Green'),
            ),
          ),
          FractionallySizedBox(
            widthFactor: 0.1,
            heightFactor: 0.1,
            alignment: Alignment.bottomLeft,
            child: Container(
              color: bottomLeftColor,
              alignment: Alignment.center,
              child: DecoratedText('Blue'),
            ),
          ),
          FractionallySizedBox(
            widthFactor: 0.1,
            heightFactor: 0.1,
            alignment: Alignment.bottomRight,
            child: Container(
              color: bottomRightColor,
              alignment: Alignment.center,
              child: DecoratedText('Red'),
            ),
          ),
          FractionallySizedBox(
            widthFactor: 0.8,
            heightFactor: 0.1,
            alignment: Alignment.topCenter,
            child: Container(
              decoration: BoxDecoration(
                  gradient: LinearGradient(colors: [
                topLeftColor,
                topRightColor,
              ])),
              child: DecoratedText('Linear Gradient'),
              alignment: Alignment.center,
            ),
          ),
          FractionallySizedBox(
            widthFactor: 0.8,
            heightFactor: 0.1,
            alignment: Alignment.bottomCenter,
            child: Container(
              decoration: BoxDecoration(
                  gradient: LinearGradient(colors: [
                bottomLeftColor,
                bottomRightColor,
              ])),
              child: DecoratedText('Linear Gradient'),
              alignment: Alignment.center,
            ),
          ),
          FractionallySizedBox(
            widthFactor: 0.1,
            heightFactor: 0.8,
            alignment: Alignment.centerLeft,
            child: Container(
              decoration: BoxDecoration(
                  gradient: LinearGradient(
                colors: [
                  topLeftColor,
                  bottomLeftColor,
                ],
                begin: Alignment.topCenter,
                end: Alignment.bottomCenter,
              )),
              child: DecoratedText('Linear Gradient'),
              alignment: Alignment.center,
            ),
          ),
          FractionallySizedBox(
            widthFactor: 0.1,
            heightFactor: 0.8,
            alignment: Alignment.centerRight,
            child: Container(
              decoration: BoxDecoration(
                  gradient: LinearGradient(
                colors: [
                  topRightColor,
                  bottomRightColor,
                ],
                begin: Alignment.topCenter,
                end: Alignment.bottomCenter,
              )),
              child: DecoratedText('Linear Gradient'),
              alignment: Alignment.center,
            ),
          ),
          FractionallySizedBox(
            widthFactor: 0.8,
            heightFactor: 0.8,
            alignment: Alignment.center,
            child: BilinearGradientDecorator(
              topLeftColor: topLeftColor,
              topRightColor: topRightColor,
              bottomLeftColor: bottomLeftColor,
              bottomRightColor: bottomRightColor,
              child: Center(child: DecoratedText('Bilinear Gradient')),
            ),
          ),
        ],
      ),
    );
  }
}

class DecoratedText extends StatelessWidget {
  const DecoratedText(this.colorName, {super.key});

  final String colorName;

  @override
  Widget build(BuildContext context) {
    return Container(
      color: Colors.black45,
      child: Text(colorName),
    );
  }
}
